Pluggable metrics collection

elevran opened the pull request:
Implementation of pluggable metrics collection.

The PR can be reviewed in three parts:
- mostly refactored and simplified metrics code from `pkg/epp/backend/metrics` now under `pkg/epp/datalayer/metrics`. Tests were moved along with the refactored code.
- go routine is no longer part of `metrics.PodMetrics` and was moved into `datalayer/collector.go` 
- changes to go.mod/sum

PR open to collect feedback. The following are yet to be done and shall be added (to this PR or a follow up)
- additional unit tests (will be added on this PR)
- logging in functions (see open question regarding context.Context in `DataSource.Collect` and `extractor.Extract`)
- hooking pluggable metrics with rest of system (follow up PR)
  - `datastore` (e.g., inferencepool target port changes, management of Collector for every endpoint, etc.)
  - default registration of the metrics data source (e.g., in `runner.go`?)
  - removal of `pkg/epp/backend/metrics`
- global statistics (and their logging) for metrics data source (separate PR or this)
- support for k8s based collectors and extractor (separate PR. Plan: extend DS interface GKV(), add new Registry for k8s based sources. Do not pass k8s sources to Collector and invoke their `Collect(ep)` directly from datastore)

elevran commented:
side note:
any idea why `fatcontext` errors were not showing up in `make lint` but are in `pull-gateway-api-inference-extension-verify-main`?
Seems to be a tighter loop if both pass/fail the same.

elevran commented:
> side note:
> any idea why fatcontext errors were not showing up in make lint but are in pull-gateway-api-inference-extension-verify-main?
> Seems to be a tighter loop if both pass/fail the same.

So apparently I was running with the latest 1.x (`1.64.8`) whereas the GIE Makefile downloads an earlier release (`1.62.2`).
Any objections to upgrading? 
I know the `fatcontext` linter was wrong in its assessment.

nirrozenbaum commented:
> So apparently I was running with the latest 1.x (`1.64.8`) whereas the GIE Makefile downloads an earlier release (`1.62.2`). Any objections to upgrading? I know the `fatcontext` linter was wrong in its assessment.

any downside from upgrading to latest (v2.3.0)?

elevran commented:
> any downside from upgrading to latest (v2.3.0)?

Not much.
v2 uses a slightly different configuration format (e.g., there's no timeout). You can convert using `golangci-lint migrate`.
The use of newer versions of linters could potentially uncover some issues which have not been reported on before, but hopefully not too many and those reported should be fixed anyway.
The only potential additional work might have to do with CICD tooling.

Suggest moving this discussion over to a new issue #1240.

nirrozenbaum commented on pkg/epp/datalayer/collector.go:
```diff
@@ -0,0 +1,121 @@
+}
+
+// Start initiates data source collection for the endpoint.
+func (c *Collector) Start(tick Ticker, ep Endpoint, registry *DataSourceRegistry) {
```
nit:
data sources are static, so it might be better to pass the sources instead of the registry.

nirrozenbaum commented on pkg/epp/datalayer/collector.go:
```diff
@@ -0,0 +1,121 @@
+// Collector runs the data collection for a single endpoint.
+type Collector struct {
+	// per-endpoint context and cancellation
+	ticker Ticker
```
if you agree that no need for ticker interface, then we don't need ticker field in the struct nor as function arg.
we can create ticker inside the start loop.  
example here:
https://github.com/kubernetes-sigs/gateway-api-inference-extension/blob/main/pkg/epp/backend/metrics/pod_metrics.go#L93

elevran replied:
the caller (e.g., datastore should own the ticker interval (or else provide it as input so we can create the timer internally).
If created internally, it *should* to be stored on the object so it's `Stop()` method can be called (this is not strictly required from Go 1.23 onwards as GC can collect timers who are no longer referenced). An explicit `Stop()` might still be preferred (e.g., to avoid ticks after `Stop` has been called?).

nirrozenbaum commented on pkg/epp/datalayer/metrics/client.go:
```diff
@@ -0,0 +1,135 @@
+	"time"
+
+	"github.com/prometheus/common/expfmt"
+	"sigs.k8s.io/gateway-api-inference-extension/pkg/epp/datalayer"
```
nit: can you split GIE imports from github imports?

nirrozenbaum commented on pkg/epp/datalayer/metrics/client.go:
```diff
@@ -0,0 +1,135 @@
+}
+
+// -- package implementations --
+var (
+	cleanupTick          = 30 * time.Second
+	maxIdleTime          = time.Minute
+	defaultClientFactory = newClientFactory()
+)
```
prefer to avoid global vars.
in this case I suggest converting tick and time to consts and removal of `defaultClientFcatory` var.
instead do:
```go
// GetDefaultClientFactory returns a default implementation of the
// ClientFactory, which caches and reuses client across calls.
func GetDefaultClientFactory() ClientFactory {
	return newClientFactory()
}
```

I understand that this might have been implemented this way to allow calling `GetDefaultClientFactory` multiple times while getting back the same variable.  I think this should be called once and stored in the caller instead.

for example, in previous version of metrics factory, here is its creation:
https://github.com/kubernetes-sigs/gateway-api-inference-extension/blob/main/cmd/epp/runner/runner.go#L265

then the same pmf variable in being used in multiple places.

elevran replied:
The use case is to allow independent callers to call GetDefaultClientFactory and get the same object back (i.e., a singleton). The case of the `pmf` (or a datastore for that matter) assumes some central point of control that coordinates everything and "injects" the `pmf` (or `datastore`) where it might be needed. This makes independent extensions more difficult (they must all be known to the coordinator). Contrast that with `GetDataStore()` that can be called where needed and retrieves the "global" data store.

Open to suggestions.

kfswain replied:
> The use case is to allow independent callers

What were you picturing the independent callers to be? I

nirrozenbaum commented on pkg/epp/datalayer/metrics/client.go:
```diff
@@ -0,0 +1,135 @@
+
+type clientmap struct {
+	cleanupOnce sync.Once
+	clients     sync.Map // map[string]*client (uses client IP as key)
```
nit: can you describe in the comment more clearly what types are used here?
https://github.com/kubernetes-sigs/gateway-api-inference-extension/blob/main/pkg/epp/datastore/datastore.go#L96

nirrozenbaum commented on pkg/epp/datalayer/metrics/client.go:
```diff
@@ -0,0 +1,135 @@
+	id := ep.GetIPAddress()
+
+	if value, loaded := clm.clients.LoadOrStore(id, cl); !loaded {
+		cl = value.(*client) // let it panic if need be
```
panic is a big no no.
always prefer to return an error.

elevran replied:
agree, but "it should never panic" (c).
we store `*client` into it so nothing else should come out.

kfswain replied:
Why not just capture the second return val (a bool) and log if its hit? That ensures graceful handling & provides a paper trail to an on-caller/SRE for debugging.

nirrozenbaum commented on pkg/epp/datalayer/metrics/client.go:
```diff
@@ -0,0 +1,135 @@
+		return nil, err
+	}
+	return metricFamilies, err
+}
```
are we supposed to update the lastUsed here?

nirrozenbaum commented on pkg/epp/datalayer/metrics/client.go:
```diff
@@ -0,0 +1,135 @@
+}
+
+func (clm *clientmap) GetClientForEndpoint(ep datalayer.Addressable) Client {
+	cl := newClient()
+	id := ep.GetIPAddress()
+
+	if value, loaded := clm.clients.LoadOrStore(id, cl); !loaded {
```
the creation of newClient for each endpoint every 50 ms seems a bit problematic.
GC might take time and memory may increase more and more with unused client objects (every cycle a new client object is created and then not used cause you read from the sync.Map).
it might be better to first check if it exists in the map before creation (even though it may add few code liens)

nirrozenbaum commented on pkg/epp/datalayer/metrics/client.go:
```diff
@@ -0,0 +1,135 @@
+
+func newClientFactory() *clientmap {
+	clm := &clientmap{}
+	clm.startCleanupGoroutine()
```
nit: do we really need the cleanupOnce?
can't we do just:
```go
clm := &clientmap{}
go clm.startCleanup() // (without cleanupOnce, without go func() inside)
...
```

elevran replied:
it goes back to the use of newClientFactory. If we can always guarantee it'll be called once, then - yes, the suggestion works.
The use of the var above and the once here was to guard against multiple calls to GetDefaultClientFactory.

nirrozenbaum commented on pkg/epp/datalayer/metrics/datasource.go:
```diff
@@ -0,0 +1,116 @@
+	if clf == nil {
+		clf = GetDefaultClientFactory()
+	}
+	ds := &DataSource{
```
nit: specific in this case we might prefer using explicit names (e.g., dataSource or dataSrc).
ds seems a bit overloaded term (used when using Datastore)

nirrozenbaum commented on pkg/epp/datalayer/metrics/client.go:
```diff
@@ -0,0 +1,135 @@
+	if err != nil {
+		return nil, err
+	}
+	return metricFamilies, err
```
metricFamilies is prometheus representation of the data, right?
do you expect every extractor to parse it to an easier format?
what are your thoughts about converting this into (key, value) pairs and allow extractor to extract a subset of (key,value) pairs?
so the conversion from metricFamilies to (key,value) pairs is done only once?

elevran replied:
The code runs in the Collector, not every extractor. You can look at the extractor's expected input type.
Can make this easier to see by adding an output type to the Collector's interface (the only reason would be "document" the output type - the code uses it when an Extractor is registered so caller access to output type is not really needed).

nirrozenbaum replied:
> I can create some helper functions that can be imported and reused, but let's do that when we write the first non-standard extractor? So we have concrete knowledge of what to reuse and how to expose it. Makes sense?

yes 👍

elevran commented on pkg/epp/datalayer/collector.go:
```diff
@@ -0,0 +1,121 @@
+}
+
+// NewTimeTicker returns a new time.Ticker with the configured duration.
+func NewTimeTicker(d time.Duration) Ticker {
```
In case I need to have control over timing of collection events (probably more in integration tests, than unit tests of the actual `Collect()` calls).
Tests could have a Mock TimeTicker that exposes a Tick() function to advance time under our test's control.

elevran replied:
Used in collector_test.go

elevran commented on pkg/epp/datalayer/metrics/extractor.go:
```diff
@@ -0,0 +1,147 @@
+}
+
+// Name returns the name of the metrics.Extractor.
+func (ext *Extractor) Name() string {
```
it is a key in the list of extractors. Should be unique and can be used to set extended attributes and trace hem back to the extractor.

elevran commented on pkg/epp/datalayer/collector.go:
```diff
@@ -0,0 +1,121 @@
+				case <-c.ticker.Channel():
+					for _, src := range sources {
+						// TODO: track errors, add context input and error return?
+						src.Collect(endpoint)
```
yes, in two ways:
- the context passed to (in future, see comment) that can be cancelled from outside; and
- collector to set timeout on the metrics` http client.
Do you think we need a way to enforce a timeout besides cancellation? That is, call with a derived context that has a timeout on it? Do you happen to know the overhead of creating a context per call (or can I reuse the context?)

kfswain commented on pkg/epp/datalayer/metrics/client.go:
```diff
@@ -0,0 +1,135 @@
+func newClient() *client {
+	return &client{
+		cl: &http.Client{
+			Timeout: 10 * time.Second,
```
Simple sounds good for now, but I bet future PRs we will need to think about configurability.

elevran replied:
the client timeout would be governed by the context passed into Get. I expect that since we scrape on a short cycle, we'll set them very low and not sure we really need to have it exposed to users. 
Other customizations (e.g., TLSConfig) would be added. Haven't thought about what needs specific to the data source (as opposed to system wide) configurations, but it would be supported via the config file, as we do for Plugins.

nirrozenbaum replied:
for the metrics client you're probably right about the short interval (not much sense in having large scrape interval).
is this true also for other collectors?
I think ideally we would like to generalize (while not over complicating) the scrape interval and timeout (which doesn't seem to be over complication).
as a starting point we could also define the timeout as a function of the scraping interval, e.g., `timeout = 3 * scrape interval`.

elevran replied:
There are two classes of http.Client timeouts:
- `http.Client.Timeout` can be used to set a hard timeout for the global client interactions (e.g., DNS resolution, TCP connection, etc.).
- per "transaction" timeouts are handled by `context.WithTimeout` and `http.NewRequestWithContext`

The initial protection was to ensure we don't have long timeouts in making connections to a non-existent server endpoint (i.e., wrong port configured, client Pod has terminated, etc.) - so main concern was the global client interaction.

Regardless, I will add "client configuration" as parameter (e.g., control over Transport.MaxIdle connections and their timeouts, TLSConfig, etc.) so there's one function to set defaults and a clear way to configure alternate values in the future should it be needed.

kfswain commented on pkg/epp/datalayer/metrics/client.go:
```diff
@@ -0,0 +1,135 @@
+	}
+}
+
+func (cl *client) Get(ctx context.Context, target *url.URL, ep datalayer.Addressable) (PrometheusMetricMap, error) {
```
Also similar to the conversation below; thats fine if we don't tackle this now

kfswain commented on pkg/epp/datalayer/metrics/extractor.go:
```diff
@@ -0,0 +1,147 @@
+		ep.UpdateMetrics(clone)
+	}
+
+	// Optional: return errors.Join(errs...) if needed
```
Yes please. Future oncallers & ops folks will thank you.

kfswain commented on pkg/epp/datalayer/metrics/datasource.go:
```diff
@@ -0,0 +1,116 @@
+	}
+	ds.extractors.Range(func(_, val any) bool {
+		if ex, ok := val.(Extractor); ok {
+			ex.Extract(families, ep) // TODO: provide context.Context, track errors?
```
Yeah, errors preferred to propagate.

kfswain commented on pkg/epp/datalayer/metrics/datasource.go:
```diff
@@ -0,0 +1,116 @@
+// DataSource is the metrics data source, returning Prometheus formatted
+// metrics from an endpoint.
+type DataSource struct {
+	mspScheme string
```
model server protocol!! I knew it was model server, cant believe i didnt put that together.

kfswain commented on pkg/epp/datalayer/metrics/datasource.go:
```diff
@@ -0,0 +1,116 @@
+
+func (ds *DataSource) getMetricsEndpoint(ep datalayer.Addressable) *url.URL {
+	return &url.URL{
+		Scheme: ds.mspScheme,
```
seeing its usage, I'm not sure `msp` is the right acronym here

kfswain commented on pkg/epp/datalayer/metrics/spec.go:
```diff
@@ -0,0 +1,207 @@
+
+// LoRASpec extends the standard Spec to allow special case
+// handling for retrieving the latest value.
+type LoRASpec struct {
```
consider extracting this to its own file, intermixing the function definition of the two implementations is challenging

elevran replied:
extracted to loraspec.go

kfswain commented on pkg/epp/datalayer/metrics/spec.go:
```diff
@@ -0,0 +1,207 @@
+
+	// Be liberal in accepting inputs that are missing quotes around label values,
+	// allowing both {label=value} and {label=\"value\"} inputs
+	quoted := addQuotesToLabelValues(spec)
```
should we be liberal in acceptance? I'm not sold on the value of accepting the performance hit of running regex in production vs just validating that metrics are working as expected in lower envs first.

elevran replied:
rationale:
- unlike k8s labels, Prometheus label values allow spaces (e.g., `status="200 OK`), so we should be prepared to handle those as well.
- standardizing on the Prometheus format allows us to use their official parser (which replace the use of the handcrafted parser code we currently have).
- this is only done *once* when the metrcis.Spec is created and not in any performance critical code. The result of parsing is stored on the spec object (name, labels) for use - just as in the current code.

So I think this change is desirable from a functional standpoint and should not have any noticeable runtime performance hit in production.
Makes sense?

elevran commented:
Addressed most of the review comments. The following are still open and I will handle over the next couple of days
- address comments regarding use of metrics.client and metrics.clientmap (e.g., global vars, client configuration, etc.)
- additional unit tests in datalayer/metrics
- some missing logs to allow debug/observe via logging messages

At this point I will ask for another review cycle. The remaining items will be handled in separate PR
- hook this into the data layer (do we want a feature flag for some soak test time or do we feel confident to switch over with existing end-to-end tests for conformance and confirmation)
- enable producing and logging cross model server metrics for the entire datastore
- additional collectors/extractors (extractor possibly needed for wide EP - leader metrics include rank so no longer in MSP expected format, workers are not exposing metrics? k8s annotations needed to determine leader/workers roles)

ahg-g commented on pkg/epp/datalayer/metrics/client.go:
```diff
@@ -0,0 +1,146 @@
+	})
+}
+
+func (clm *clientmap) GetClientForEndpoint(ep datalayer.Addressable) (Client, error) {
```
The whole complexity in this file is because we are creating a client per endpoint, why do we need to do that? currently we use a single client: https://github.com/kubernetes-sigs/gateway-api-inference-extension/blob/main/cmd/epp/runner/runner.go#L267

is there a scalability issue with that?

elevran replied:
Potentially. The http.Client has a limit on MaxIdleConnections (default: 100). The concern is that as the number of scraped targets goes over that limit, the connections would be closed and them immediately reopened on the next cycle to that target. This is not handled to the best of my understanding in the current code. 
The implication is increased latency (1 RTT per request/response without the limit vs 3 RTT - socket, TLS, request-response), operating system overheads in opening/closing connections, and potential descriptor exhaustion (sockets going into timed-wait might not be available for reuse in time). 

Client per target avoids that and caps the number of sockets in the system to O(targets).
Another, much simpler option, would be to set MaxIdleConnection to a high enough number (1000?). Is that a reasonable starting point? It would eliminate client caching wholly from the code. We can then iterate should it become an issue (we should track Get's latency for that).

elevran replied:
implemented with high maxIdleConns. We can revisit in future PR.

elevran commented on pkg/epp/datalayer/collector.go:
```diff
@@ -0,0 +1,131 @@
+}
+
+// Start initiates data source collection for the endpoint.
+func (c *Collector) Start(ctx context.Context, ticker Ticker, ep Endpoint, sources []DataSource) error {
```
Simplifies testing, IMO, to allow passing in a mock.Ticker and advance time as fast as needed without having to rely on go routine scheduling assumptions and longer than necessary sleep's.
What is the downside?

ahg-g replied:
keeping it as is is fine

elevran commented on pkg/epp/datalayer/collector.go:
```diff
@@ -0,0 +1,131 @@
+					for _, src := range sources {
+						ctx, cancel := context.WithTimeout(c.ctx, defaultCollectionTimeout)
+						_ = src.Collect(ctx, endpoint) // TODO: track errors per collector
+						cancel()                       // release the ctx timeout resources
```
Context would be GC'ed but I don't think its cancel shall be called. I don't know if the GC is sufficient to release all resources in this case or an explicit cancel is needed.

elevran commented on pkg/epp/datalayer/endpoint.go:
```diff
@@ -38,3 +41,60 @@ type Endpoint interface {
 	AttributeMap
 }
+
+// ModelServer is an implementation of the Endpoint interface.
```
Used the same structure as `PodMetrics` - an `atomic.Pointer` to `Pod` info (not the k8s Pod). So I think there "Pod" (IP, name, labels) per "ModelServer" but a Pod could potentially appear in multiple ModelServers.

elevran commented:
@ahg-g anything else that should be addressed in this PR?

kfswain commented:
/lgtm
/approve
